### PR TITLE
JVM_IR: when generating invoke(), take it from the expected fn type

### PR DESCRIPTION
**Pull request by pyos**
#KT-47483 Fixed

### CODE REVIEW

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AssertionLowering.kt [hunk 1]
```diff
@@ -77,39 +76,23 @@ private class AssertionLowering(private val context: JvmBackendContext) :
...
-
-            return if (mode == JVMAssertionsMode.ALWAYS_ENABLE) {
-                checkAssertion(assertCondition, lambdaArgument)
-            } else {
-                require(mode == JVMAssertionsMode.JVM && data != null)
-                irIfThen(
-                    irNot(getAssertionDisabled(this, data)),
-                    checkAssertion(assertCondition, lambdaArgument)
-                )
+        val constructor = context.ir.symbols.assertionErrorConstructor
```
Comment (dnpetrov) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AssertionLowering.kt:79]: How are these changes related to KT-47483?

**Comment by dnpetrov** [comment]
Please, add box tests.

**Comment by MoroIddrisu** [comment]
Hello

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AssertionLowering.kt (hunk 1 above)
Comment (MoroIddrisu) [reply compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/AssertionLowering.kt:79]: Hello

**Comment by dnpetrov** [comment]
@pyos what's up with this PR? Are you still working on it? Should it be discarded?

**Comment by MoroIddrisu** [comment]
Hello

**Comment by dnpetrov** [comment]
@pyos ping.
